contracts/Vault.sol

```diff
@@ -123,21 +130,27 @@ contract Vault is ExchangeRate, Initializable, OwnableUpgradeSafe {
             aTokenAmount = balance;
         }
 
+        require(aTokenAmount > 0, "amount to withdraw cannot be zero");
+        bToken.burn(address(this), balance);
+        require(aTokenAmount >= depositorToPrincipal[msg.sender], "atoken must be more than depositorPrincipal");
+        require(depositorReserve >= SafeMath.sub(aTokenAmount, depositorToPrincipal[msg.sender]), "atoken must be more than depositorPrincipal");
         depositorReserve = SafeMath.sub(
-            depositorReserve, 
+            depositorReserve,
             SafeMath.sub(aTokenAmount, depositorToPrincipal[msg.sender])
             );
+        require(principal >= depositorToPrincipal[msg.sender], "principal must be over depositorPrincipal");
         principal = SafeMath.sub(principal, depositorToPrincipal[msg.sender]);
         depositorToPrincipal[msg.sender] = 0;
         // withdraw atokens
-        bToken.burn(address(this), balance);
         aToken.approve(address(aavePool), aTokenAmount);
-        aavePool.withdraw(
+        uint256 withdrawn = aavePool.withdraw(
             address(token),
             aTokenAmount,
             msg.sender
         );
-        _updateCheckpointInterest();
+        if (startedSurplus) {
+            _updateCheckpointInterest();
+        }
         emit DepositorWithdraw(msg.sender, aTokenAmount);
     }
```

Comment:
is this function safe from the multiple withdraws attack: https://www.researchgate.net/publication/334161350_Resolving_the_Multiple_Withdrawal_Attack_on_ERC20_Tokens?

Reply:
am not too familiar with this particular scenario. From reading just the abstract, the attack involves a malicious contract front running withdrawal(s)

our contract should be safe, since we:
1. withdraw an entire balance of the user, hence the multiple withdrawal is not achievable 
2. explicitly moving the bTokens from msg.sender, hence not callable by third party

interesting attacks though, will add this into my reading list